### ISSUE TITLE
aws ecs register-task-definition created, but after update service tasks can't be run

### ISSUE THREAD

**Issue by mRevuelta** [aws/aws-cli#3887]
To whom it may concern:

First of all thanks in advance for reading this issue.

Note that the launch type is EC2. The amazon AMI being used by the EC2 instance in the cluster is amzn-ami-2018.03.j-amazon-ecs-optimized (ami-0de29b072b458b107).

aws-cli version used is aws-cli/1.16.93 Python/2.7.10 Darwin/17.7.0 botocore/1.12.83

If you need further details about infrastructure/set up just tell me.

So, I've been able to create task definitions and services through the aws web console, which end in a task running in my ecs cluster correctly. 

But when I create a new revision for the same task definition (or any new task definition) through the aws cli (configured with the same I AM user which worked successfully with the web console), when the service tries to run the task it fails 

I've notice that when I create a new task definition via web console two required attributes are appended to the task definition, but that never happens when I try to create task definitions or new task definition revisions via aws cli (tried with two different ones, one running local and another downloaded to a CircleCi job):

I've been reviewing the docs and tried different ways to workaround this issue with no success. 

task-definition-revision-1 via web console, everything went fine, here's the description:
```json{
  "ipcMode": null,
  "executionRoleArn": "arn:aws:iam::XXXXXXXXXXX:role/ecsTaskExecutionRole",
  "containerDefinitions": [
    {
      "dnsSearchDomains": null,
      "logConfiguration": null,
      "entryPoint": null,
      "portMappings": [
        {
          "hostPort": 8080,
          "protocol": "tcp",
          "containerPort": 8080
        }
      ],
      "command": null,
      "linuxParameters": null,
      "cpu": 1024,
      "environment": [],
      "ulimits": null,
      "dnsServers": null,
      "mountPoints": [],
      "workingDirectory": null,
      "secrets": null,
      "dockerSecurityOptions": null,
      "memory": 768,
      "memoryReservation": null,
      "volumesFrom": [],
      "image": "XXXXXXXXXXX.dkr.ecr.eu-west-1.amazonaws.com/gateway:dev-3.4.0-SNAPSHOT-build-14609",
      "disableNetworking": null,
      "interactive": null,
      "healthCheck": null,
      "essential": true,
      "links": null,
      "hostname": null,
      "extraHosts": null,
      "pseudoTerminal": null,
      "user": null,
      "readonlyRootFilesystem": null,
      "dockerLabels": null,
      "systemControls": null,
      "privileged": null,
      "name": "gateway-container"
    }
  ],
  "placementConstraints": [],
  "memory": "1024",
  "taskRoleArn": null,
  "compatibilities": [
    "EC2"
  ],
  "taskDefinitionArn": "arn:aws:ecs:eu-west-1:XXXXXXXXXXX:task-definition/poc-taskdef-gateway:1",
  "family": "poc-taskdef-gateway",
  "requiresAttributes": [
    {
      "targetId": null,
      "targetType": null,
      "value": null,
      "name": "ecs.capability.execution-role-ecr-pull"
    },
    {
      "targetId": null,
      "targetType": null,
      "value": null,
      "name": "com.amazonaws.ecs.capability.ecr-auth"
    }
  ],
  "pidMode": null,
  "requiresCompatibilities": [
    "EC2"
  ],
  "networkMode": null,
  "cpu": "1024",
  "revision": 1,
  "status": "ACTIVE",
  "volumes": []
}
```

Note in the previous block that after creating the task definition through the web console, the required attributes *ecr-auth and *ecr-pull are there. I've even checked the http post payload that generates the web console and its the same that I used with the cli.

But, when I create a new revision for that task definition through the aws cli (same credentials as the user working on the web console), the output has no requiredAttributes and despite the task definition new revision is created and the service updated, the service seems to be unable to run the task:

task-definition-revision-2:
```json{
  "ipcMode": null,
  "executionRoleArn": "arn:aws:iam::XXXXXXXXXXX:role/ecsTaskExecutionRole",
  "containerDefinitions": [
    {
      "dnsSearchDomains": null,
      "logConfiguration": null,
      "entryPoint": null,
      "portMappings": [
        {
          "hostPort": 8080,
          "protocol": "tcp",
          "containerPort": 8080
        }
      ],
      "command": null,
      "linuxParameters": null,
      "cpu": 1024,
      "environment": [],
      "ulimits": null,
      "dnsServers": null,
      "mountPoints": [],
      "workingDirectory": null,
      "secrets": null,
      "dockerSecurityOptions": null,
      "memory": 768,
      "memoryReservation": null,
      "volumesFrom": [],
      "image": "XXXXXXXXXXX.dkr.ecr.eu-west-1.amazonaws.com/gateway:dev-3.4.0-SNAPSHOT-build-14609",
      "disableNetworking": null,
      "interactive": null,
      "healthCheck": null,
      "essential": true,
      "links": null,
      "hostname": null,
      "extraHosts": null,
      "pseudoTerminal": null,
      "user": null,
      "readonlyRootFilesystem": null,
      "dockerLabels": null,
      "systemControls": null,
      "privileged": null,
      "name": "gateway-container"
    }
  ],
  "placementConstraints": [],
  "memory": null,
  "taskRoleArn": null,
  "compatibilities": [
    "EC2"
  ],
  "taskDefinitionArn": "arn:aws:ecs:eu-west-1:XXXXXXXXXXX:task-definition/poc-taskdef-gateway:2",
  "family": "poc-taskdef-gateway",
  "requiresAttributes": [],
  "pidMode": null,
  "requiresCompatibilities": [],
  "networkMode": null,
  "cpu": null,
  "revision": 2,
  "status": "ACTIVE",
  "volumes": []
}
```

Apart that the required attributes are missing, I can see that im obtaining the following error: 

Status reason | CannotPullContainerError: Error response from daemon: Get https://XXX-dkr.ecr.eu-west-1.amazonaws.com/v2/: dial tcp: lookup XXXXXXXXXXX-dkr.ecr.eu-west-1.amazonaws.com on XXX.XX.X.X:XX: no such host

and at the ecs.log I can see this one:

[WARN] Managed task [arn:aws:ecs:eu-west-1:XXXXXXXXXXX:task/312ff28a-57cf-4073-87f0-82a0d48bdb40]: error creating container [gateway-container]; marking its desired status as STOPPED: Error: No such image: XXXXXXXXXXX.dkr.ecr.eu-west-1.amazonaws.com/gateway:dev-3.4.0-SNAPSHOT-build-14609

I would really appreciate any hint that helps me to solve this situation, I would really like to use the aws cli to manage the ECS cluster. 

What I'm doing wrong? Did I missed setting up some configuration in the aws cli?

I'm a bit out of ideas after some hours, any suggestion is welcome.

Thanks!

**Comment by justnance** [comment]
@mRevuelta - Thank you for reaching out.  The details provided were really helpful.  It appears there are a couple of issues here.  

First, the `CannotPullContainerError` and  `error creating container` usually returns when there is a networking issue.   As per the API documentation, [If you are using the EC2 launch type, any network mode can be used. If the network mode is set to none, you cannot specify port mappings in your container definitions, and the tasks containers do not have external connectivity. ](https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_RegisterTaskDefinition.html)
 
Also note, the [Container instances need external network access to communicate with the Amazon ECS service endpoint, so if your container instances do not have public IP addresses, then they must use network address translation (NAT) to provide this access.](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/launch_container_instance.html)

Another option would be to enable the subnet to [auto-assign public IPs](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-ip-addressing.html#subnet-public-ip). 

Regarding the "required attributes are missing", I will have to run a few tests and reproduce this behavior before I can confirm if the issue is with the CLI itself, the service, or our documentation requires modification.  I will update the issue again shortly.

**Comment by mRevuelta** [comment]
@justnance - Thanks for your fast reply and suggestions. 

I'm setting up a new cluster to run in a VPC and subnet with the suggested set up, in parallel, I'll also try specifying the network mode.

 I'll keep the issue updated with whatever happens.

Thanks again!

**Comment by mRevuelta** [comment]
Hi again @justnance, specifying the network mode to bridge solved the issue 👍 

If this is a strange behaviour and you want more information about the circumstances that lead to this issue, feel free to tell me.

Thanks once again for your time and have a nice day :)

**Comment by justnance** [comment]
@mRevuelta - Thank you for your feedback and update.  I'm closing the issue.
